### PR TITLE
[enhancement] Seperate API server support

### PR DESCRIPTION
**Pull request by atb00ker**
Checks:

- [x] I have manually tested the proposed changes
- [x] I have written new test cases to avoid regressions (if necessary)
- [x] I have updated the documentation (e.g. README.rst)

Blocked https://github.com/openwisp/openwisp-utils/pull/210
Tested In: https://github.com/openwisp/docker-openwisp/pull/154

### CODE REVIEW

File: openwisp_monitoring/device/admin.py [hunk 1]
```diff
@@ -127,7 +129,13 @@ def get_extra_context(self, pk=None):
...
         if pk:
             device_data = DeviceData(pk=uuid.UUID(pk))
-            api_url = reverse('monitoring:api_device_metric', args=[pk])
+            api_url = reverse(
+                'monitoring:api_device_metric',
+                urlconf=MONITORING_API_URLCONF,
+                args=[pk],
+            )
+            if MONITORING_API_BASEURL:
+                api_url = urljoin(MONITORING_API_BASEURL, api_url)
```
Comment (atb00ker) [inline openwisp_monitoring/device/admin.py:138]: Notable change!

File: openwisp_monitoring/device/apps.py [hunk 2]
```diff
@@ -201,7 +205,20 @@ def register_dashboard_items(self):
...
+            )
+            device_list_url = reverse_lazy(
+                "monitoring:api_location_device_list",
+                urlconf=MONITORING_API_URLCONF,
+                args=["000"],
+            )
+            if MONITORING_API_BASEURL:
+                device_list_url = urljoin(MONITORING_API_BASEURL, str(device_list_url))
+                loc_geojson_url = urljoin(MONITORING_API_BASEURL, str(loc_geojson_url))
+
```
Comment (atb00ker) [inline openwisp_monitoring/device/apps.py:221]: Notable change!

File: openwisp_monitoring/device/apps.py [hunk 3]
```diff
@@ -218,4 +235,8 @@ def register_dashboard_items(self):
                         'monitoring/js/leaflet.fullscreen.min.js',
                     ),
                 },
+                extra_config={
+                    'monitoring_device_list_url': device_list_url,
+                    'monitoring_location_geojson_url': loc_geojson_url,
+                },
```
Comment (atb00ker) [inline openwisp_monitoring/device/apps.py:241]: Notable change!

File: openwisp_monitoring/device/templates/admin/dashboard/device_map.html [hunk 4]
```diff
@@ -1,9 +1,9 @@
 {% load leaflet_tags i18n %}
 <script type="text/javascript">
-window._owGeoMapConfig = {
-  geoJsonUrl: '{% url "monitoring:api_location_geojson" %}',
-  locationDeviceUrl: '{% url "monitoring:api_location_device_list" "000" %}?page_size=5'
-}
+  window._owGeoMapConfig = {
+    geoJsonUrl: '{{ monitoring_location_geojson_url }}',
+    locationDeviceUrl: '{{ monitoring_device_list_url }}?page_size=5'
```
Comment (atb00ker) [inline openwisp_monitoring/device/templates/admin/dashboard/device_map.html:5]: Notable change!

**Review by atb00ker — COMMENTED** [review]
Too much noise in this PR because of linters.
Marked the notable changes! :smile:

**Comment by atb00ker** [comment]
>  I have written new test cases to avoid regressions (if necessary)

Not sure how to test this one; to change the values of `DASHBOARD_TEMPLATE`, I will somehow need to restart the django test server which might cause other problems in parallel testing.

File: openwisp_monitoring/device/tests/test_admin.py [hunk 5]
```diff
@@ -37,17 +37,25 @@ def test_device_admin(self):
...
+        self.assertContains(response, 'http://testserver/api')
+        self.assertContains(response, check.name)
+        self.assertContains(response, check.params)
+
+    def test_dashboard_map_on_index(self):
+        url = reverse('admin:index')
+        self._login_admin()
+        response = self.client.get(url)
+        self.assertContains(response, "geoJsonUrl: \'http://testserver/api")
+        self.assertContains(response, "locationDeviceUrl: \'http://testserver/api")
```
Comment (atb00ker) [inline openwisp_monitoring/device/tests/test_admin.py:58]: Notable change!

File: requirements.txt [hunk 6]
```diff
@@ -1,6 +1,7 @@
 # TODO: put back the standard pypi release when
 # next point version of openwisp-controller is released
 openwisp-controller @ https://github.com/openwisp/openwisp-controller/tarball/master
+openwisp-utils[rest] @ https://github.com/atb00ker/openwisp-utils/tarball/extraconfig
```
Comment (atb00ker) [inline requirements.txt:4]: Notable change!

File: tests/openwisp2/settings.py [hunk 7]
```diff
@@ -190,6 +190,8 @@
 # avoid slowing down the test suite with mac vendor lookups
 if TESTING:
     OPENWISP_MONITORING_MAC_VENDOR_DETECTION = False
+    OPENWISP_MONITORING_API_URLCONF = 'openwisp_monitoring.urls'
+    OPENWISP_MONITORING_API_BASEURL = 'http://testserver'
```
Comment (atb00ker) [inline tests/openwisp2/settings.py:194]: Notable change!

File: openwisp_monitoring/device/static/monitoring/js/device-map.js [hunk 8]
```diff
@@ -179,5 +180,12 @@ ${pagination}
         }
 
         loadingOverlay.fadeOut(250);
+    }
+
+    $.ajax({
+        dataType: "json",
+        url: _owGeoMapConfig.geoJsonUrl,
+        xhrFields: { withCredentials: true },
+        success: onAjaxSuccess
```
Comment (atb00ker) [inline openwisp_monitoring/device/static/monitoring/js/device-map.js:189]: Noteable change!

File: openwisp_monitoring/device/static/monitoring/css/percircle.css [hunk 9]
```diff
@@ -1 +1 @@
...
... 0 .1em .1em 0;background-color: #ccc}.percircle *,.percircle :after,.percircle :before{box-sizing: content-box}.percircle.animate:after,.percircle.animate>span{-webkit-transition: -webkit-transform .2s ease-in-out;transition: transform .2s ease-in-out}.percircle.animate .bar{-webkit-transition: -webkit-transform .6s ease-in-out;transition: transform .6s ease-in-out}.percircle.center{float: none;margin: 0 auto}.percircle.big{font-size: 240px}.percircle.small{font-size: 80px}.percircle>span{position: absolute;z-index: 1;width: 100%;top: 50%;top: calc(50% - .1em);transform: translateY(-50%);height: 1em;font-size: .2em;color: #ccc;display: block;text-align: center;white-space: nowrap}.perclock>span{font-size: .175em}.percircle:after{position: absolute;top: .08em;left: .08em;display: block;content: " ";border-radius: 50%;background-color: #f5f5f5;width: .84em;height: .84em}.percircle .slice{position: absolute;width: 1em;height: 1em;clip: rect(0,1em,1em,.5em)}.percircle:hover{cursor: default}.percircle:hover>span{-webkit-transform: scale(1.3) translateY(-50%);transform: scale(1.3) translateY(-50%);color: #307bbb}.percircle:hover:after{-webkit-transform: scale(1.1);transform: scale(1.1)}
```
Comment (nemesisdesign) [inline openwisp_monitoring/device/static/monitoring/css/percircle.css:1]: what's going on here?

**Review by nemesisdesign — COMMENTED** [review]
@atb00ker it would have been better to separate the linting changes into a dedicate commit.

I don't see changes in README, did you forget to add these two settings there?

File: requirements-test.txt [hunk 10]
```diff
@@ -1,5 +1,6 @@
 # unpinned, the version is dictated by openwisp-controller
-openwisp-utils[qa]
+# TODO: Remove before release
+openwisp-utils[qa] @ https://github.com/atb00ker/openwisp-utils/tarball/extraconfig
```
Comment (atb00ker) [inline requirements-test.txt:3]: Need to change before merge.

File: requirements.txt [hunk 11]
```diff
@@ -1,8 +1,8 @@
 # TODO: put back the standard pypi release when
 # next point version of openwisp-controller is released
 openwisp-controller @ https://github.com/openwisp/openwisp-controller/tarball/master
-# openwisp-utils[rest] is left unpinned so the version used
-# is the one dictated by openwisp-controller
+# TODO: Remove before release
```
Comment (atb00ker) [inline requirements.txt:4]: Need to change before merge.

**Review by atb00ker — COMMENTED** [review]
> @atb00ker it would have been better to separate the linting changes into a dedicate commit.

Yes, done that, hopefully now it will be clearer.

> I don't see changes in README, did you forget to add these two settings there?

https://github.com/openwisp/openwisp-monitoring/pull/304/files#diff-7b3ed02bc73dc06b7db906cf97aa91dec2b2eb21f2d92bc5caa761df5bbc168fR816-R817

> what's going on here?

The file is okay, github seems to be having issues showing it properly because of width in compare view.
This file only contains linting changes.

File: requirements.txt (hunk 11 above)
Comment (nemesisdesign) [reply requirements.txt:4]: I think this was merged and you should be able to change this, right?

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

File: requirements.txt [hunk 12]
```diff
@@ -1,8 +1,7 @@
 # TODO: put back the standard pypi release when
 # next point version of openwisp-controller is released
 openwisp-controller @ https://github.com/openwisp/openwisp-controller/tarball/master
-# openwisp-utils[rest] is left unpinned so the version used
-# is the one dictated by openwisp-controller
+openwisp-utils[rest] @ https://github.com/openwisp/openwisp-utils/tarball/master
```
Comment (nemesisdesign) [inline requirements.txt:4]: this shouldn't be needed: https://github.com/openwisp/openwisp-controller/blob/master/requirements.txt#L10

Comment (atb00ker) [reply requirements.txt]: removed!

**Review by nemesisdesign — APPROVED** [review]
Thanks :+1:

**Pull request merged**